File: scala/src/test/scala/scalasearch/SearchResultTest.scala
package scalasearch

import org.junit.Assert._
import org.junit.runner.RunWith
import org.scalatest.FunSuite
import org.scalatest.junit.JUnitRunner

@RunWith(classOf[JUnitRunner])
class SearchResultTest extends FunSuite {
  test("test single-line search result") {
    val pattern = "Search".r
    val searchFile = new SearchFile("~/src/git/xsearch/csharp/CsSearch/CsSearch",
      "Searcher.cs", FileType.Text)
    val lineNum = 10
    val matchStartIndex = 15
    val matchEndIndex = 23
    val line = "\tpublic class Searcher\n"
    val searchResult = new SearchResult(pattern, searchFile, lineNum,
      matchStartIndex, matchEndIndex, line)
    val expectedPath = "~/src/git/xsearch/csharp/CsSearch/CsSearch/Searcher.cs"
    val expectedOutput = "%s: %d: [%d:%d]: %s".format(expectedPath,
      lineNum, matchStartIndex, matchEndIndex, line.trim)
    System.out.println("\nsearchResult.toString():\n" + searchResult.toString)
    System.out.println("expectedOutput:\n" + expectedOutput)
    assertEquals(searchResult.toString, expectedOutput)
  }

  test("test binary file search result") {
    val pattern = "Search".r
    val searchFile = new SearchFile("~/src/git/xsearch/csharp/CsSearch/CsSearch",
      "Searcher.exe", FileType.Binary)
    val lineNum = 0
    val matchStartIndex = 0
    val matchEndIndex = 0
    val line:String = null
    val searchResult = new SearchResult(pattern, searchFile, lineNum,
      matchStartIndex, matchEndIndex, line)
    val expectedPath = "~/src/git/xsearch/csharp/CsSearch/CsSearch/Searcher.exe"
    val expectedOutput = "%s matches".format(expectedPath)
    System.out.println("\nsearchResult.toString():\n" + searchResult.toString)
    System.out.println("expectedOutput:\n" + expectedOutput)
    assertEquals(searchResult.toString, expectedOutput)
  }

  test("test multi-line search result") {
    val pattern = "Search".r
    val searchFile = new SearchFile("~/src/git/xsearch/csharp/CsSearch/CsSearch",
      "Searcher.cs", FileType.Text)
    val lineNum = 10
    val matchStartIndex = 15
    val matchEndIndex = 23
    val line = "\tpublic class Searcher\n"
    val linesBefore = List("namespace CsSearch\n", "{\n")
    val linesAfter = List("\t{\n", "\t\tprivate readonly FileTypes _fileTypes;\n")
    val searchResult = new SearchResult(pattern, searchFile, lineNum,
      matchStartIndex, matchEndIndex, line, linesBefore, linesAfter)
    val expectedPath = "~/src/git/xsearch/csharp/CsSearch/CsSearch/Searcher.cs"
    val expectedOutput =
      """================================================================================
        |%s: %d: [%d:%d]
        |--------------------------------------------------------------------------------
        |   8 | namespace CsSearch
        |   9 | {
        |> 10 | 	public class Searcher
        |  11 | 	{
        |  12 | 		private readonly FileTypes _fileTypes;
        |""".stripMargin.format(expectedPath, lineNum, matchStartIndex, matchEndIndex)
    System.out.println("\nsearchResult.toString():\n" + searchResult.toString)
    System.out.println("expectedOutput:\n" + expectedOutput)
    assertEquals(searchResult.toString, expectedOutput)
  }
}

File: node/tests/searchresult_test.js
/*
 * searchresult_test.js
 *
 * Some nodeunit tests of searchresult.js
 */

var common = require('../nodesearch/common.js');
var SearchResult = require('../nodesearch/searchresult.js').SearchResult;

exports.testSingleLineResult = function(test) {
    var pattern = 'Search';
    var file = '~/src/git/xsearch/csharp/CsSearch/CsSearch/Searcher.cs';
    var linenum = 10;
    var matchStartIndex = 15;
    var matchEndIndex = 23;
    var line = "\tpublic class Searcher\n";
    var linesBefore = [];
    var linesAfter = [];
    var result = new SearchResult(pattern, file, linenum, matchStartIndex,
        matchEndIndex, line, linesBefore, linesAfter);
    var resultString = result.toString();
    var expectedOutput = file + ": " + linenum + ": [" + matchStartIndex + ":" +
        matchEndIndex + "]: " + line.trim();
    test.ok(resultString === expectedOutput, "single-line result matches expected");
    test.done();
}

exports.testBinaryFileResult = function(test) {
    var pattern = 'Search';
    var file = '~/src/git/xsearch/csharp/CsSearch/CsSearch/Searcher.exe';
    var linenum = 0;
    var matchStartIndex = 0;
    var matchEndIndex = 0;
    var line = "";
    var linesBefore = [];
    var linesAfter = [];
    var result = new SearchResult(pattern, file, linenum, matchStartIndex,
        matchEndIndex, line, linesBefore, linesAfter);
    var resultString = result.toString();
    var expectedOutput = file + " matches";
    test.ok(resultString === expectedOutput, "single-line result matches expected");
    test.done();
}

exports.testMultiLineResult = function(test) {
    var pattern = 'Search';
    var file = '~/src/git/xsearch/csharp/CsSearch/CsSearch/Searcher.cs';
    var linenum = 10;
    var matchStartIndex = 15;
    var matchEndIndex = 23;
    var line = "\tpublic class Searcher\n";
    var linesBefore = ["namespace CsSearch\n", "{\n"];
    var linesAfter = ["\t{\n", "\t\tprivate readonly FileTypes _fileTypes;\n"];

    var result = new SearchResult(pattern, file, linenum, matchStartIndex,
        matchEndIndex, line, linesBefore, linesAfter);
    var resultString = result.toString();
    var outputTemplate = "" +
        "================================================================================\n" +
        "{0}: {1}: [{2}:{3}]\n" +
        "--------------------------------------------------------------------------------\n" +
        "   8 | namespace CsSearch\n" +
        "   9 | {\n" +
        "> 10 | \tpublic class Searcher\n" +
        "  11 | \t{\n" +
        "  12 | \t\tprivate readonly FileTypes _fileTypes;\n";
    var expectedOutput = outputTemplate.format(file, linenum,
        matchStartIndex, matchEndIndex);
    test.ok(resultString === expectedOutput, "multi-line result matches expected");
    test.done();
}
